### ISSUE TITLE
Rules to link accounts by email are case-sensitive

### ISSUE THREAD

**Issue by alanc10n** [auth0/rules#208]
I've encountered an issue in which a user signs up with multiple connections associated with the same email address and the automatic account linking rule fails to link them. Eventually I tracked this down to the fact that the `users-by-email` endpoint performs a case-sensitive search, which is very surprising. The user signed up for an email/password account as `user@example.com`, but their Google account has `USER@EXAMPLE.COM` as the associated email address. 

In practice, email addresses are treated in a case-insensitive manner, and the Auth0 web console performs case-insensitive search for email addresses. It seems like anyone using these rules to link accounts with matching emails would want it handle differing case, since some providers will report differing case for a given email address (i.e. not everyone normalizes addresses to lower-case).

**Comment by rhiadj** [comment]
I was caught out by this over the past week and even Auth0 support seemed flummoxed by this and unaware that the email address used to add admins to a tenant was case sensitive.

**_Some info:_**

http://www.faqs.org/rfcs/rfc821.html
> For some hosts the user name is case sensitive, and SMTP implementations must take case to preserve the case of user names as they appear in mailbox arguments.  Host names are not case sensitive.

However...
http://www.faqs.org/rfcs/rfc5321.html
> However, exploiting the  case sensitivity of mailbox local-parts impedes interoperability and is discouraged.

And as always Stack Overflow to the rescue:
https://stackoverflow.com/questions/9807909/are-email-addresses-case-sensitive

So it appears it depends on the RFC you are implementing.

RFC-5321 is the latest (RFC-7504 is only an ‘update’ to it) so I'd suggest this should be the one to follow - which suggests _*case insensitivity*_ should be used.

Discussing this with other (seasoned) developers - 99% of devs I've spoken to were unaware that email addresses can be case sensitive. And in practice most systems implementing SMTP do so using case insensitive implementation.

**_Suggested/Possible Solutions:_**

Either change this to be case insensitive as suggested in the most recent RFC

_... or..._

At the very least the fields/processes/APIs used to register and to add admins to a tenant should state the email address is case sensitive. And error message(s) displayed should detail this fact also.

Thanks.

**Comment by rhiadj** [comment]
Just to add to this - the existing error message I was getting when attempting to log in states:
![auth0_error_message](https://user-images.githubusercontent.com/1793567/68850953-dd0e0500-06cc-11ea-9423-9ecbc51a0d8f.jpg)

... which is exactly what I did!

In Auth0 the invite was created using:
`Rhiad.Jaffar@emailaddress.com`

I received the email invite in my inbox for:
`rhiad.jaffar@emailaddress.com`
... because the email server is set up to ignore case sensitivity.

So the error message above is in fact incorrect (or at the very least misleading) for around 90% of use cases where case sensitivity is to blame due to the fact that most email servers are set up to be case insensitive?

Thanks.

**Comment by rhiadj** [comment]
@alanc10n Thanks for raising this issue as I wondered if I was going mad but evidently wasn't the only one - which was good to know! 👍 

It might be worth checking this again just fyi.

I reported this issue to customer support who have since informed me that the dev team had pushed out a fix for this and that user registration is now case insensitive.

I've not had chance to check myself yet but thought I'd pass on the message in case it's blocking your work/endeavours!

Support also mentioned they'll pass on this issue for the dev team to respond to.

Thanks again.

**Comment by sebhs** [comment]
Just spend hours trying to debug our merging process and realized that this was the issue. The user signed up via Google with the email `Johndoe@gmail.com` and then passwordless with `johndoe@gmail.com`, because of the case sensitivity of the `users-by-email` endpoint, the user wasn't merged properly. 

Does anyone have a good solution how to resolve this? From the last comment it seems like this should have been fixed, but the we had this problem yesterday.

**Comment by Maxhodges** [comment]
Auth0 know about this issue since 2019 and hasn't been fixed yet? For Christ's sake!

We just lost a customer because of a nightmare scenario involving this bug. Please make it right! We just got the worst possible review from a customer who has a second account created because of a capitalization difference which too a week of back and forth for our customer support and engineers to figure out.

> 0 out of 10 
> the site has a glitch where if you change your email on your account, it force-makes accounts instead of logging in. I logged into my account several times and several times it just made a new account instead. This led to confusion from the staff as they thought I was hitting the 'getting started' instead of signing in, which I had to take several screenshots to prove I wasn't until I was finally taken seriously and the staff removed the other accounts and troubleshot the original. This took a week to fix with a bit of frustration from having the back and forth of 'stop making new accounts'/'I'm not making new accounts, the site is'.

**Comment by alanc10n** [comment]
> Hi everyone. I had the same issue using /users-by-email endpoint because is case-sensitive (it doesn't make sense). I found the https://{domain}/api/v2/users endpoint allows you to search by different attributes such as email. So doing the following query I was able to find the users with case-insensitive string.
> 
> `?q=email%3Auser_email` which decoded is `?q=email:user_email`
> 
> Please let me know if this works for you.

Yes, that search is case-insensitive, but it's also not immediately consistent. [Auth0 explicitly recommends using users-by-email for searching during authentication and account-linking](https://auth0.com/docs/manage-users/user-search/retrieve-users-with-get-users-by-email-endpoint), which is why this is a mess.

**Comment by casperdcl** [comment]
Can we have a new case-insensitive endpoint option? Users frequently capitalise the first letter of their email *sometimes but not always*, and *expect* the backend to deduplicate.

Until auth0 is fixed, I'm forced to implement some work-around dedupe logic post-login:

```py
def email_norm(s: str):
    s = s.lower()
    s = re.sub(r'@googlemail\.com$', '@gmail.com', s)
    if re.search(r'@gmail\.com$', s):
        s = re.sub(r'\.', '', s.split('@', 1)[0]) + '@gmail.com' # remove dots
        s = re.sub(r'\+.*@', '@', s)                             # remove +labels
    return s

def find_email(email: str, db: Iterable[str]):
    norm = email_norm(email)
    return next((e for e in db if norm == email_norm(e)), None)

# check auth0 login_email against internal DB
if match := find_email(login_email, database_emails):
    if match != login_email:
        raise ValueError(
            f"You have attempted to login as `{login_email}` but `{match}` already exists."
            f"Please logout and use `{match}` instead.")
else:
    database_emails.add(login_email) # deliberately doesn't use email_norm
```

Though silly issues like this make me feel like switching away from auth0... perhaps even self-implementing a solution from scratch using things like https://github.com/pyauth/pyotp
